data_model/src/metadata.rs

```diff
@@ -0,0 +1,324 @@
+        for k in path.iter().take(path.len() - 1) {
+            map = match map.get(k)? {
+                Value::LimitedMetadata(data) => &data.map,
+                _ => return None,
+            };
+        }
+        map.get(key)
+    }
+
+    /// Remove leaf node in metadata, given path. If the path is
+    /// malformed, or incorrect (if e.g. any of interior path segments
+    /// are not [`Metadata`] instances) return `None`. Else return the
+    /// owned value corresponding to that path.
+    pub fn nested_remove(&mut self, path: &Path) -> Option<Value> {
+        let key = path.last()?;
+        let mut map = &mut self.map;
+        for k in path.iter().take(path.len() - 1) {
+            map = match map.get_mut(k)? {
+                Value::LimitedMetadata(data) => &mut data.map,
+                _ => return None,
+            };
+        }
+        map.remove(key)
+    }
+
+    /// Insert the given [`Value`] into the given path. If the path is
+    /// complete, check the limits and only then insert. The creation
+    /// of the path is the responsibility of the user.
+    ///
+    /// # Errors
+    /// - If the path is empty.
+    /// - If one of the intermediate keys is absent.
+    /// - If some intermediate key is a leaf node.
+    pub fn nested_insert_with_limits(
+        &mut self,
+        path: &Path,
+        value: Value,
+        limits: Limits,
+    ) -> Result<Option<Value>> {
+        if self.map.len() >= limits.max_len as usize {
+            return Err(eyre!(
+                "Metadata length limit is reached: {}",
+                limits.max_len
+            ));
+        }
+        let key = path.last().ok_or_else(|| eyre!("Empty path"))?;
+        let mut layer = self;
+        for k in path.iter().take(path.len() - 1) {
+            layer = match layer
+                .map
+                .get_mut(k)
+                .ok_or_else(|| eyre!("No metadata for key {} in path. Path is malformed.", k))?
+            {
+                Value::LimitedMetadata(data) => data,
+                _ => return Err(eyre!("Path contains non-metadata segments at key {}.", k)),
+            };
+        }
+        check_size_limits(key, value.clone(), limits)?;
+        layer.insert_with_limits(key.to_string(), value, limits)
+    }
+
+    /// Insert [`Value`] under the given key.  Returns `Some(value)`
+    /// if the value was already present, `None` otherwise.
+    ///
+    /// # Errors
+    /// Fails if `max_entry_byte_size` or `max_len` from `limits` are exceeded.
+    pub fn insert_with_limits(
+        &mut self,
+        key: Name,
+        value: Value,
+        limits: Limits,
+    ) -> Result<Option<Value>> {
+        if self.map.len() >= limits.max_len as usize && !self.map.contains_key(&key) {
+            return Err(eyre!(
+                "Metadata length limit is reached: {}",
+                limits.max_len
+            ));
+        }
+        check_size_limits(&key, value.clone(), limits)?;
```

Comment:
Though each entry size is limited whole metadata size is unlimited.
This seems to me not to represent `LimitedMetadata` concept, if I understand correctly

Reply:
Whole metadata size is limited. Remember that you can only insert one entry at a time, and each time that you do, there are checks involved. Problems could arise if I allowed things like `UnlimitedMetadata` be entries in limited metadata types, which I don't allow.